lib/constructs/interface.js

```diff
@@ -416,7 +421,83 @@ class Interface {
   }
 
   generateIterator() {
-    if (this.iterable && this.iterable.isPair) {
+    if (!this.iterable) {
+      return;
+    }
+
+    if (this.iterable.isAsync) {
+      this.str += `
+        const AsyncIteratorPrototype = Object.create(utils.AsyncIteratorPrototype, {
+          [Symbol.toStringTag]: {
+            value: "${this.name} AsyncIterator",
+            configurable: true
+          }
+        });
+
+        Object.assign(AsyncIteratorPrototype, {
+          next() {
+            const internal = this[utils.iterInternalSymbol];
+            if (!internal) {
+              return Promise.reject(new TypeError("next() called on an object that is not an async iterator prototype object"));
+            }
+
+            const nextSteps = () => {
+              if (internal.isFinished) {
+                return Promise.resolve({ value: undefined, done: true });
+              }
+
+              const nextPromise = internal.target[implSymbol][utils.asyncIteratorNext](this);
+              return nextPromise.then(next => {
+                internal.ongoingPromise = null;
+                if (next === utils.asyncIteratorEOI) {
+                  internal.isFinished = true;
+                  return { value: undefined, done: true };
+                }`;
+      if (this.iterable.isPair) {
+        this.str += `
+                return utils.iteratorResult(next.map(utils.tryWrapperForImpl), kind);
+        `;
+      } else {
+        this.str += `
+                return { value: utils.tryWrapperForImpl(next), done: false };
+        `;
+      }
+      this.str += `
+              });
+            };
+
+            internal.ongoingPromise = internal.ongoingPromise ? internal.ongoingPromise.then(nextSteps) : nextSteps();
+            return internal.ongoingPromise;
+          }
+      `;
+
+      if (this.iterable.hasReturnSteps) {
```

Comment:
What do you mean by "have return added in the generated file"?

Reply:
I don't think that works because we only require the `Impl` class at the bottom of the file, I think for circular dependency reasons. At which time the `implementation` export might not exist.